### PR TITLE
Implement subscription stop position

### PR DESCRIPTION
**Pull request by alexrudd**
Continuing work from: https://github.com/liftbridge-io/liftbridge-api/pull/44

Opening as a draft as could do with some guidance.

To implement `StopPosition_STOP_TIMESTAMP` I'd need to find the latest offset _before_ the specified timestamp. The CommitLog interface only exposes [`OffsetForTimestamp()`](https://github.com/liftbridge-io/liftbridge/blob/master/server/commitlog/interface.go#L27) which returns the earliest offset _after_ a timestamp.

Looks like I'd have to alter this interface to expose either:

```go
NearestOffsetBeforeTimestamp(timestamp int64) (int64, error)
NearestOffsetAfterTimestamp(timestamp int64) (int64, error)
```

or

```go
NearestOffsetToTimestamp(timestamp int64, before bool) (int64, error)
```

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -705,6 +705,37 @@ func getStartOffset(req *client.SubscribeRequest, log commitlog.CommitLog) (int6
...
+		var err error
+		stopOffset, err = log.OffsetForTimestamp(req.StopTimestamp)
+		if err != nil {
+			return stopOffset, status.New(
+				codes.Internal, fmt.Sprintf("Failed to lookup offset for timestamp: %v", err))
+		}
+	case client.StopPosition_STOP_LATEST:
+		stopOffset = log.NewestOffset()
+		if stopOffset == -1 {
+			return stopOffset, status.New(codes.ResourceExhausted, "Stream is empty %s")
```
Comment (tylertreat) [inline server/api.go:728]: Status message has a string format character in it.

**Review by tylertreat — COMMENTED** [review]
lgtm. I'm good with altering the commitlog interface to:

```go
NearestOffsetBeforeTimestamp(timestamp int64) (int64, error)
NearestOffsetAfterTimestamp(timestamp int64) (int64, error)
```

File: server/api_test.go [hunk 2]
```diff
@@ -842,6 +842,103 @@ func TestSubscribePartitionClosed(t *testing.T) {
...
+		// we've recieved all messages so this must be a resource exhaused error
+		if receivedCount > expectedCount {
+			require.Error(t, err)
+			require.Equal(t, lift.ErrStreamExhausted, err)
+			exhausted <- struct{}{}
+		}
+	},
+		lift.StartAtTime(start.CommitTimestamp()),
+		lift.StopAtTime(stop.CommitTimestamp()),
+	)
```
Comment (alexrudd) [inline server/api_test.go:932]: This test is failing unexpectedly, which makes me think I've misunderstood something fundamental.

The start time and stop time options should be range inclusive, so setting them to the known values of when two messages were committed to the log, should return a stream of messages that includes those two messages at the beginning and end of the stream respectively (offsets 2-6). Instead the first messages delivered by the stream is offset 3.

File: server/commitlog/commitlog.go [hunk 3]
```diff
@@ -328,6 +328,55 @@ func (l *commitLog) OffsetForTimestamp(timestamp int64) (int64, error) {
 	return l.segments[len(l.segments)-1].NextOffset(), nil
 }
 
+// LatestOffsetBeforeTimestamp returns the latest offset whose timestamp is less
+// than or equal to the given timestamp.
+func (l *commitLog) LatestOffsetBeforeTimestamp(timestamp int64) (int64, error) {
```
Comment (alexrudd) [inline server/commitlog/commitlog.go:333]: There's probably a much more efficient/concise way of implementing this, but I couldn't see it tonight

File: server/api_test.go [hunk 4]
```diff
@@ -842,6 +842,103 @@ func TestSubscribePartitionClosed(t *testing.T) {
...
+
+	stream := "foo"
+	ctx := context.Background()
+	err = client.CreateStream(ctx, "foo", stream)
+	require.NoError(t, err)
+
+	// subscription to empty stream errors
+	err = client.Subscribe(ctx, stream, nil, lift.StopAtLatestReceived())
+	require.Error(t, err)
+	require.Equal(t, lift.ErrStreamExhausted, err)
```
Comment (alexrudd) [inline server/api_test.go:871]: Locally I've added support for stop position in go-liftbridge. One change I made was to introduce a new sentinal error: `ErrStreamExhausted` for when the grpc code `ResourceExhausted` is returned. This meant changing the API slightly, as previous a `ErrReadonlyPartition` would be returned for the code `ResourceExhausted`.

Adding the new error might not actually be necessary though?

**Comment by alexrudd** [comment]
Finally got a bit more time to work on this. The most recent code seems to work, though not exactly as I was expecting (see [this comment](https://github.com/liftbridge-io/liftbridge/pull/286#discussion_r523293293)). Would appreciate any feedback.

I'll take a fresh look at this tomorrow and see if I can improve things.

File: server/commitlog/commitlog.go (hunk 3 above)
Comment (alexrudd) [reply server/commitlog/commitlog.go:333]: Think I've tidied this up somewhat

File: server/commitlog/segment.go [hunk 5]
```diff
@@ -383,22 +383,22 @@ func (s *segment) findEntry(offset int64) (e *entry, err error) {
...
+		if err := s.Index.ReadEntryAtLogOffset(e, int64(i)); err != nil {
 			panic(err)
 		}
 		return e.Timestamp >= timestamp
 	})
 	if idx == n {
 		return nil, ErrEntryNotFound
 	}
-	err = s.Index.ReadEntryAtFileOffset(e, int64(idx*entryWidth))
+	err = s.Index.ReadEntryAtLogOffset(e, int64(idx))
```
Comment (alexrudd) [inline server/commitlog/segment.go:401]: Making use of the `ReadEntryAtLogOffset` and `CountEntries` helpers here seemed neater.

**Review by tylertreat — COMMENTED** [review]
Overall changes look good. I don't love the idea of having to create a new v3 of go-liftbridge for the error change, but the readonly error feels weird to return.

File: server/api_test.go (hunk 2 above)
Comment (alexrudd) [reply server/api_test.go:932]: Yep that worked, thanks!

File: server/api.go [hunk 6]
```diff
@@ -590,9 +590,14 @@ func (a *apiServer) subscribe(ctx context.Context, partition *partition,
...
-	if partition.IsReadonly() {
-		stopOffset = partition.log.NewestOffset()
+	stopOffset, st := getStopOffset(req, partition.log)
+	if st != nil {
+		return nil, nil, st
+	}
+
+	if stopOffset != waitForNewMessages && stopOffset < startOffset {
+		return nil, nil, status.New(
+			codes.Internal, fmt.Sprintf("Stop offset is before start offset: %d < %d", stopOffset, startOffset))
```
[the remark this answers is not in the document]
Comment (alexrudd) [reply server/api.go:600]: Yeah that makes more sense, I'm not too familiar with grpc codes

File: server/commitlog/commitlog.go [hunk 7]
```diff
@@ -328,6 +328,52 @@ func (l *commitLog) OffsetForTimestamp(timestamp int64) (int64, error) {
...
+	}
+	// Search the previous segment for the first entry whose timestamp is
+	// greater than or equal to the given timestamp. If this is the first
+	// segment, just search it.
+	var seg *segment
+	if idx == 0 {
+		seg = l.segments[0]
+		// if the given timestamp is before the start of the stream return an
+		// error.
+		if timestamp < seg.firstWriteTime {
```
[the remark this answers is not in the document]
Comment (alexrudd) [reply server/commitlog/commitlog.go:351]: Have added a method to retrieve `firstWriteTime` with read locking

**Comment by alexrudd** [comment]
Think this is ready for review now. It depends on https://github.com/liftbridge-io/go-liftbridge/pull/96 being merged though for the tests to pass.

File: server/api.go [hunk 8]
```diff
@@ -602,7 +607,7 @@ func (a *apiServer) subscribe(ctx context.Context, partition *partition,
 	)
 	if err != nil {
 		return nil, nil, status.New(
-			codes.Internal, fmt.Sprintf("Failed to create stream reader: %v", err))
+			codes.InvalidArgument, fmt.Sprintf("Failed to create stream reader: %v", err))
```
Comment (tylertreat) [inline server/api.go:610]: I think you changed the code for the wrong error. `Internal` is correct here and line 600 should be an `InvalidArgument` IMO.

**Review by tylertreat — CHANGES_REQUESTED** [review]
One minor issue with the error codes—looks like the wrong one got swapped.

I've also merged the go-liftbridge PR, so that can be updated here.

File: server/api.go (hunk 8 above)
Comment (alexrudd) [reply server/api.go:610]: ah, good spot

**Pull request merged**